feat(auth): Auth advanced workflows for JS and RN (#6185)
aws-amplify/docs: src/fragments/lib/auth/js/advanced.mdx

## Identity Pool Federation

You can alternatively create your own custom credentials provider to get AWS credentials directly from Cognito Federated Identities and not use User Pool federation. You must supply the custom credentials provider to Amplify via the `Amplify.configure` method call. Below, you can see sample code of how such a custom provider can be built to achieve the use case. 

```js
import { Amplify } from "aws-amplify";
import {
  fetchAuthSession,
  AWSCredentialsAndIdentityIdProvider,
  AWSCredentialsAndIdentityId,

>   TokenProvider,

Reviewer: You dont need TokenProvider

  AuthTokens,
} from "aws-amplify/auth";

> import { CognitoIdentity } from "@aws-sdk/client-cognito-identity";

Reviewer: I would add a comment that customers need to install `@aws-sdk/client-cognito-identity` package

// You can make use of the sdk to get identityId and credentials
const cognitoidentity = new CognitoIdentity({
  region: "<region-from-config>",
});

// Note: The custom provider class must implement AWSCredentialsAndIdentityIdProvider
class CustomCredentialsProvider implements AWSCredentialsAndIdentityIdProvider {

  // Example class member that holds the login information 
  federatedLogin?: {
    domain: string,
    token: string
  };

  // Custom method to load the federated login information
  loadFederatedLogin(login?: typeof this.federatedLogin) { 
    // You may also persist this by caching if needed

>     this.federatedLogin = login;

Reviewer: What if instead of expecting a value, you pass a callback function that returns `federatedLogin` this way when `getCredentialsAndIdentityId` is invoke it will get the latest information?

> 

Reviewer 2: ## Subscribing **to** Events
You can take specific actions when users sign-in or sign-out by subscribing **to** authentication events in your app. Please see our [Hub Module Developer Guide](/lib/utilities/hub) for more information.

## Identity Pool Federation

You can alternatively create your own custom credentials provider to get AWS credentials directly from Cognito Federated Identities and not use User Pool federation. You must supply the custom credentials provider to Amplify via the `Amplify.configure` method call. Below, you can see sample code of how such a custom provider can be built to achieve the use case. 

```js

> import { Amplify } from "aws-amplify";
> import {
>   fetchAuthSession,
>   AWSCredentialsAndIdentityIdProvider,
>   AWSCredentialsAndIdentityId,
>   GetCredentialsOptions,
>   AuthTokens,
> } from "aws-amplify/auth";
> 
> // Note: This example requires installing `@aws-sdk/client-cognito-identity` to obtain Cognito credentials
> // npm i @aws-sdk/client-cognito-identity
> import { CognitoIdentity } from "@aws-sdk/client-cognito-identity";
> 
> // You can make use of the sdk to get identityId and credentials
> const cognitoidentity = new CognitoIdentity({
>   region: "<region-from-config>",
> });

Reviewer 2: Nit:
```suggestion
import { Amplify } from 'aws-amplify';
import {
  fetchAuthSession,
  AWSCredentialsAndIdentityIdProvider,
  AWSCredentialsAndIdentityId,
  GetCredentialsOptions,
  AuthTokens,
} from 'aws-amplify/auth';

// Note: This example requires installing `@aws-sdk/client-cognito-identity` to obtain Cognito credentials
// npm i @aws-sdk/client-cognito-identity
import { CognitoIdentity } from '@aws-sdk/client-cognito-identity';

// You can make use of the sdk to get identityId and credentials
const cognitoidentity = new CognitoIdentity({
  region: '<region-from-config>',
});
```

// Note: The custom provider class must implement AWSCredentialsAndIdentityIdProvider
class CustomCredentialsProvider implements AWSCredentialsAndIdentityIdProvider {

  // Example class member that holds the login information 
  federatedLogin?: {
    domain: string,
    token: string
  };

  // Custom method to load the federated login information
  loadFederatedLogin(login?: typeof this.federatedLogin) { 
    // You may also persist this by caching if needed
    this.federatedLogin = login;
  }

  async getCredentialsAndIdentityId(
    getCredentialsOptions: GetCredentialsOptions
  ): Promise<AWSCredentialsAndIdentityId | undefined> {
    try {

      // You can add in some validation to check if the token is available before proceeding
      // You can also refresh the token if it's expired before proceeding

      const getIdResult = await cognitoidentity.getId({
        // Get the identityPoolId from config

>         IdentityPoolId: "<identity-pool-id-from-config>",

Reviewer 2: Nit:
```suggestion
        IdentityPoolId: '<identity-pool-id-from-config>',
```

        Logins: { [this.federatedLogin.domain]: this.federatedLogin.token },
      });

      const cognitoCredentialsResult = await cognitoidentity.getCredentialsForIdentity({
        IdentityId: getIdResult.IdentityId,
        Logins: { [this.federatedLogin.domain]: this.federatedLogin.token },
      });

      const credentials: AWSCredentialsAndIdentityId = {
        credentials: {
          accessKeyId: cognitoCredentialsResult.Credentials?.AccessKeyId,
          secretAccessKey: cognitoCredentialsResult.Credentials?.SecretKey,
          sessionToken: cognitoCredentialsResult.Credentials?.SessionToken,
          expiration: cognitoCredentialsResult.Credentials?.Expiration,
        },
        identityId: getIdResult.IdentityId,
      };
      return credentials;
    } catch (e) {

>       console.log("Error getting credentials: ", e);

Reviewer 2: Nit:
```suggestion
      console.log('Error getting credentials: ', e);
```

    }
  }
  // Implement this to clear any cached credentials and identityId. This can be called when signing out of the federation service.
  clearCredentialsAndIdentityId(): void {} 
}

// Create an instance of your custom provider
const customCredentialsProvider = new CustomCredentialsProvider(); 
Amplify.configure(awsconfig, {
  Auth: {
    // Supply the custom credentials provider to Amplify
    credentialsProvider: customCredentialsProvider
  },
});

```

Now that the custom credentials provider is built and supplied to `Amplify.configure`, let's look at how you can use the custom credentials provider to finish federation into Cognito identity pool.

<InlineFilter filters = {['react-native']}>

### Facebook Sign-in (React Native - Expo)

```javascript
import Expo from 'expo';
import React from 'react';
import {
  fetchAuthSession,

> } from "aws-amplify/auth";

Reviewer 2: Nit:
```suggestion
} from 'aws-amplify/auth';
```

const App = () => {
  const signIn = async () => {
    const { type, token, expires } = await Expo.Facebook.logInWithReadPermissionsAsync('YOUR_FACEBOOK_APP_ID', {
        permissions: ['public_profile'],
      });
    if (type === 'success') {
      // sign in with federated identity
      try {

>         customCredentialsProvider.loadFederatedLogin({
>           domain: "graph.facebook.com",
>           token: token,
>         });
>         const fetchSessionResult = await fetchAuthSession(); // will return the AWS credentials
>         console.log("fetchSessionResult: ", fetchSessionResult); 

Reviewer 2: Nit:
```suggestion
        customCredentialsProvider.loadFederatedLogin({
          domain: 'graph.facebook.com',
          token: token,
        });
        const fetchSessionResult = await fetchAuthSession(); // will return the AWS credentials
        console.log('fetchSessionResult: ', fetchSessionResult); 
```

      } catch(err) {
        console.log(err);
      }
    }
  }

  // ...

  return (
    <View style={styles.container}>
      <Button title="FBSignIn" onPress={signIn} />
    </View>
  );

}
```
</InlineFilter>

<InlineFilter filters = {['js']}>

### Facebook sign-in (React)

```js
import React, { useEffect } from 'react';
import {
  fetchAuthSession,

> } from "aws-amplify/auth";

Reviewer 2: nit: 
```suggestion
} from 'aws-amplify/auth';
```

            if (!response || !response.authResponse) {
              return;
            }
            customCredentialsProvider.loadFederatedLogin({

>               domain: "graph.facebook.com",

Reviewer 2: nit:
```suggestion
              domain: 'graph.facebook.com',
```

              token: response.authResponse.accessToken,
            });
            const fetchSessionResult = await fetchAuthSession(); // will return the AWS credentials

>             console.log("fetchSessionResult: ", fetchSessionResult); 

Reviewer 2: nit:
```suggestion
            console.log('fetchSessionResult: ', fetchSessionResult); 
```

}
```

### Google sign-in (React)

```jsx
import React, { useEffect } from 'react';
import jwt from 'jwt-decode';
import {
  fetchAuthSession,

> } from "aws-amplify/auth";

Reviewer 2: nit:
```suggestion
} from 'aws-amplify/auth';
```

      window.google.accounts.id.initialize({
        client_id: process.env.GOOGLE_CLIENT_ID,
        callback: (response: any) => {

>           customCredentialsProvider.loadFederatedLogin({
>             domain: "accounts.google.com",
>             token: response.credential,
>           });
>           const fetchSessionResult = await fetchAuthSession(); // will return the AWS credentials
>           console.log("fetchSessionResult: ", fetchSessionResult); 

Reviewer 2: nit: 
```suggestion
          customCredentialsProvider.loadFederatedLogin({
            domain: 'accounts.google.com',
            token: response.credential,
          });
          const fetchSessionResult = await fetchAuthSession(); // will return the AWS credentials
          console.log('fetchSessionResult: ', fetchSessionResult); 
```

  }

  return (
    <div>
      <button id="googleSignInButton"/>
    </div>
  );
}
```

### Federate with Auth0

You can use `Auth0` as one of the providers of your Cognito Identity Pool. This will allow users authenticated via Auth0 have access to your AWS resources.

Step 1. [Follow Auth0 integration instructions for Cognito Federated Identity Pools](https://auth0.com/docs/integrations/integrating-auth0-amazon-cognito-mobile-apps)

Step 2. Login with `Auth0`, then use the id token returned to get AWS credentials from `Cognito Federated Identity Pools` using custom credentials provider you created at the start:

```js
import {
  fetchAuthSession,
} from "aws-amplify/auth";

const { idToken, domain, name, email, phoneNumber } = getFromAuth0(); // get the user credentials and info from auth0

async function getCognitoCredentials () {
  try {
     customCredentialsProvider.loadFederatedLogin({
        domain,
        token: idToken,
      });
      const fetchSessionResult = await fetchAuthSession(); // will return the AWS credentials
      console.log("fetchSessionResult: ", fetchSessionResult); 
  } catch(err) {
    console.log(err);
  }
}
```
</InlineFilter>

## Lambda Triggers

> The CLI allows you to configure [Lambda Triggers](https://docs.aws.amazon.com/cognito/latest/developerguide/cognito-user-identity-pools-working-with-aws-lambda-triggers.html) for your Cognito user pool. These enable you to add custom functionality to your registration and authentication flows. [Read more](/cli/function)

Reviewer: ```suggestion
The CLI allows you to configure [Lambda Triggers](https://docs.aws.amazon.com/cognito/latest/developerguide/cognito-user-identity-pools-working-with-aws-lambda-triggers.html) for your Cognito User Pool. These enable you to add custom functionality to your registration and authentication flows. [Read more](/cli/function)
```

```ts
import { signIn } from 'aws-amplify/auth';

async function handleSignIn(username, password) {
  try {
    await signIn({
      username,
      password,
      options: {
        serviceOptions: {
          clientMetadata: {} // Optional, an object of key-value pairs which can contain any key and will be passed to your Lambda trigger as-is.
        }
      }
    });
  } catch(err) {
    console.log(err);
  }
}
```

### Passing metadata to other Lambda triggers

Many Cognito Lambda Triggers also accept unsanitized key/value pairs in the form of a `clientMetadata` attribute.  This attribute can be specified for various Auth APIs which result in Cognito Lambda Trigger execution.

These APIs include:

- `signIn`
- `signUp`
- `confirmSignIn`
- `confirmSignUp`
- `resetPassword`
- `confirmResetPassword`
- `resendSignUpCode`
- `updateUserAttributes`

Please note that some of triggers which accept a `validationData` attribute will use `clientMetadata` as the value for `validationData`.  Exercise caution with using `clientMetadata` when you are relying on `validationData`.

## Working with AWS service objects

> You can use AWS *Service Interface Objects* to work AWS Services in authenticated State. You can call methods on any AWS Service interface object by passing your credentials from Amplify to the service call constructor:

Reviewer: `fetchAuthSession` ?

    }
  }

  return (
    <div>
      <button id="googleSignInButton"/>
    </div>
  );
}
```

### Federate with Auth0

You can use `Auth0` as one of the providers of your Cognito Identity Pool. This will allow users authenticated via Auth0 have access to your AWS resources.

Step 1. [Follow Auth0 integration instructions for Cognito Federated Identity Pools](https://auth0.com/docs/integrations/integrating-auth0-amazon-cognito-mobile-apps)

Step 2. Login with `Auth0`, then use the id token returned to get AWS credentials from `Cognito Federated Identity Pools` using custom credentials provider you created at the start:

```js
import {
  fetchAuthSession,

> } from "aws-amplify/auth";

Reviewer 2: nit: 
```suggestion
} from 'aws-amplify/auth';
```

const { idToken, domain, name, email, phoneNumber } = getFromAuth0(); // get the user credentials and info from auth0

async function getCognitoCredentials () {
  try {
     customCredentialsProvider.loadFederatedLogin({
        domain,
        token: idToken,
      });
      const fetchSessionResult = await fetchAuthSession(); // will return the AWS credentials

>       console.log("fetchSessionResult: ", fetchSessionResult); 

Reviewer 2: nit:
```suggestion
      console.log('fetchSessionResult: ', fetchSessionResult); 
```

  } catch(err) {
    console.log(err);
  }
}
```
</InlineFilter>

## Lambda Triggers

The CLI allows you to configure [Lambda Triggers](https://docs.aws.amazon.com/cognito/latest/developerguide/cognito-user-identity-pools-working-with-aws-lambda-triggers.html) for your Cognito User Pool. These enable you to add custom functionality to your registration and authentication flows. [Read more](/cli/function)

### Pre Authentication and Pre Sign-up Lambda triggers

If you have a Pre Authentication Lambda trigger enabled, you can pass `clientMetadata` as an option for `signIn`. This metadata can be used to implement additional validations around authentication.

```ts
import { signIn } from 'aws-amplify/auth';

> async function handleSignIn(username, password) {

Reviewer 3: we should add the type definition here. If customer have `implicitAny` to true it will give them an error

  try {
    await signIn({
      username,
      password,
      options: {

>         serviceOptions: {

Reviewer 3: we removed serviceOptions already, it should be inside options
